src/ConnectedMode/Migration/VsAwareFileSystem.cs

```diff
@@ -58,52 +61,148 @@ internal class VsAwareFileSystem : IVsAwareFileSystem
             return Task.CompletedTask;
         }
 
         private async Task<IVsQueryEditQuerySave2> GetSccServiceAsync()
         {
             IVsQueryEditQuerySave2 sccService = null;
-            await threadHandling.RunOnUIThread( () =>
+            await threadHandling.RunOnUIThread(() =>
             {
                 sccService = serviceProvider.GetService(typeof(SVsQueryEditQuerySave)) as IVsQueryEditQuerySave2;
             });
 
             return sccService;
         }
+
+        private async Task CheckOutFilesToDeleteAsync(string folderPath)
+        {
+            // Can't check out the directory for saving - if you do, VS will pop up a dialogue,
+            // even if the "silent" flag is set.
+            bool success = await CheckoutForEditAsync(folderPath);
+
+            if (success)
+            {
+                var files = fileSystem.Directory.GetFiles(folderPath, "*.*", System.IO.SearchOption.AllDirectories);
+                success = await CheckoutForSaveAsync(files);
+            }
+
+            if (!success)
+            {
+                var message = string.Format(MigrationStrings.VsFileSystem_Error_FailedToCheckOutFolderForDeletion, folderPath);
+                throw new InvalidOperationException(message);
+            }
+        }
+
+        private async Task<bool> CheckoutForEditAsync(params string[] fileNames)
+        {
+            Debug.Assert(sccService != null, "sccService should not be null");
+            logger.LogMigrationVerbose("Checking out file(s) for edit: " +
+                string.Join(", ", fileNames));
+
+            VsQueryEditFlags flags = VsQueryEditFlags.SilentMode | VsQueryEditFlags.DetectAnyChangedFile |
+                // Force no prompting here to fix SLVS#801 (otherwise TFS would prompt about the slconfig
+                // and generated ruleset files in new connected mode that are not in the solution)
+                VsQueryEditFlags.ForceEdit_NoPrompting;
+
+            uint verdict = 0;
+            uint moreInfo = 0;
+
+            await threadHandling.RunOnUIThread(() =>
+            {
+                Microsoft.VisualStudio.ErrorHandler.ThrowOnFailure(sccService.QueryEditFiles((uint)flags, fileNames.Length, fileNames, null, null, out verdict, out moreInfo));
+            });
+
+            var success = (tagVSQueryEditResult.QER_EditOK == (tagVSQueryEditResult)verdict);
+            if (!success)
+            {
+                this.logger.WriteLine(MigrationStrings.VSFileSystem_FailedToCheckOutFilesForEditing, (tagVSQueryEditResultFlags)moreInfo);
+            }
+            return success;
+        }
+
+        private async Task<bool> CheckoutForSaveAsync(params string[] fileNames)
+        {
+            bool result = false;
+
+            await threadHandling.RunOnUIThread(() =>
+                result = DoCheckoutForSave(fileNames));
+
+            return result;
+        }
+
+        private bool DoCheckoutForSave(params string[] fileNames)
```

Comment:
Copied the existing [method](https://github.com/SonarSource/sonarlint-visualstudio/blob/03123f51a5f22aa2bd6e347f00a33ca01e9ee44b/src/Integration/LocalServices/SourceControlledFileSystem.cs#L158).

I didn't change this method to handle threading; it was simpler to add a thread-aware wrapper method instead.